### PR TITLE
Visject Improvements

### PR DESCRIPTION
**Pull request by stefnotch**
Addresses a part of https://github.com/FlaxEngine/FlaxAPI/issues/22
-   Arrow Keys & Enter to choose an autocomplete suggestion
-   Alternative names for nodes (`*` for multiply, `+` for add, ...)
-   Smart:tm: node connecting

-   Visject Surface keeps it's focus on node creation
-   Newly created nodes get added to the selection
-   Node gets removed from selection when you add some connections

- Various minor fixes

### CODE REVIEW

File: FlaxEditor/Surface/ContextMenu/VisjectCM.cs [hunk 1]
```diff
@@ -314,39 +314,55 @@ public override bool OnKeyDown(Keys key)
...
-            return null;
         }
 
         /// <summary>
-        /// Gets the previous visible sibling of a node
+        /// Gets the next siblings of a control that have a specific type.
+        /// </summary>
+        /// <param name="item">A control that is attached to a parent</param>
+        /// <returns>An <see cref="IEnumerable{T}"/> with the siblings that come after the current one.</returns>
+        private IEnumerable<T> GetNextSiblings<T>(Control item) where T : Control
```
Comment (stefnotch) [inline FlaxEditor/Surface/ContextMenu/VisjectCM.cs]: Would this (and the other similar methods in this file) be a good addition to Flax?

File: FlaxEditor/Surface/VisjectSurface.Serialization.cs [hunk 2]
```diff
@@ -96,109 +96,109 @@ private static void ReadCommonValue(BinaryReader stream, ref object value)
...
-                break;
-            }
-
-            default: throw new SystemException();
+            {
+                case 0: // CommonType::Bool:
+                    value = stream.ReadByte() != 0;
+                    break;
+                case 1: // CommonType::Integer:
+                    {
```
Comment (stefnotch) [inline FlaxEditor/Surface/VisjectSurface.Serialization.cs]: The standard C# Visual Studio formatter changes the formatting of lines like those.

File: FlaxEditor/Surface/ContextMenu/VisjectCM.cs [hunk 3]
```diff
@@ -314,39 +314,55 @@ public override bool OnKeyDown(Keys key)
...
             }
-            return null;
+        }
+
+        /// <summary>
+        /// Gets the previous sibling of a control that have a specific type.
+        /// </summary>
+        /// <param name="item">A control that is attached to a parent</param>
+        /// <returns>An <see cref="IEnumerable{T}"/> with the siblings that come before the current one.</returns>
+        private IEnumerable<T> GetPreviousSiblings<T>(Control item) where T : Control
```
Comment (stefnotch) [inline FlaxEditor/Surface/ContextMenu/VisjectCM.cs]: Would this (and the other similar methods in this file) be a good addition to Flax?

File: FlaxEditor/Surface/Archetypes/Math.cs [hunk 4]
```diff
@@ -103,7 +108,7 @@ private static NodeArchetype Op2(ushort id, string title, string desc, Connectio
             Op2(20, "Dot", "Returns the dot product of A and B", ConnectionType.Vector, ConnectionType.Float, false),
             Op2(21, "Max", "Selects the greater of A and B"),
             Op2(22, "Min", "Selects the lesser of A and B"),
-            Op2(23, "Power", "Returns A raised to the specified at B power"),
+            Op2(23, "Power", "Returns A raised to the specified at B power", new []{ "^", "**" }), //TODO: Fight a religious war over which operator is better
```
Comment (mafiesto4) [inline FlaxEditor/Surface/Archetypes/Math.cs]: I think `^` is better but both can stay for now.

**Pull request merged**